dev/ToastNotifications/ToastNotificationUtility.cpp

```diff
@@ -0,0 +1,212 @@
+        &hKey,
+        nullptr /* lpdwDisposition */));
+
+    std::wstring comRegistrationExeString{ c_quote + processName.get() + c_quote + c_toastActivatedArgument };
+
+    RegisterValue(hKey, nullptr, reinterpret_cast<const BYTE*>(comRegistrationExeString.c_str()), REG_SZ, (comRegistrationExeString.size() * sizeof(wchar_t)));
+}
+
+void UnRegisterComServer(std::wstring const& clsid)
+{
+    wil::unique_hkey hKey;
+    //subKey: Software\Classes\CLSID\{comActivatorGuidString}\LocalServer32
+    std::wstring subKey{ c_clsIdPath + clsid + LR"(\LocalServer32)" };
+
+    THROW_IF_FAILED(RegDeleteKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        KEY_WOW64_64KEY,
+        0));
+}
+
+void UnRegisterAppIdentifierFromRegistry()
+{
+    std::wstring appIdentifier{ RetrieveAppId() };
+
+    wil::unique_hkey hKey;
+    //subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appIdentifier };
+
+    THROW_IF_FAILED(RegDeleteKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        KEY_WOW64_64KEY,
+        0));
+}
+
+std::wstring RegisterComActivatorGuidAndAssets(winrt::Microsoft::Windows::ToastNotifications::ToastActivationInfo const& details)
+{
+    std::wstring appIdentifier{ RetrieveAppId() };
+    THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appIdentifier.c_str(), GUID_NULL));
+
+    wil::unique_hkey hKey;
+    // subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appIdentifier };
+
+    THROW_IF_FAILED(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr /* lpClass */,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr /* lpSecurityAttributes */,
+        &hKey,
+        nullptr /* lpdwDisposition */));
+
+    WCHAR registeredGuidBuffer[GUID_LENGTH + 4]; // GUID length + '{' + '}' + '\0'
+    DWORD bufferLength = sizeof(registeredGuidBuffer);
+    HRESULT status = RegGetValueW(
+        hKey.get(),
+        nullptr /* lpValue */,
+        L"CustomActivator",
+        RRF_RT_REG_SZ,
+        nullptr /* pdwType */,
+        &registeredGuidBuffer,
+        &bufferLength);
+
+    if (status == ERROR_FILE_NOT_FOUND)
+    {
+        wil::unique_cotaskmem_string processName;
+        THROW_IF_FAILED(GetCurrentProcessPath(processName));
+
+        GUID comActivatorGuid = GUID_NULL;
+        THROW_IF_FAILED(CoCreateGuid(&comActivatorGuid));
+
+        // StringFromCLSID returns GUID String with braces.
+        wil::unique_cotaskmem_string comActivatorGuidString;
+        THROW_IF_FAILED(StringFromCLSID(comActivatorGuid, &comActivatorGuidString));
+
+        RegisterAssets(appIdentifier, details.Assets(), comActivatorGuidString);
+        RegisterComServer(processName, comActivatorGuidString);
+
+        return comActivatorGuidString.get();
+    }
+    else
+    {
+        THROW_HR_IF(status, FAILED_WIN32(status));
```

Comment:
I believe we have a wil check for win32 error codes. Trying to find the right macro to use here

Reply:
THROW_IF_WIN32_ERROR